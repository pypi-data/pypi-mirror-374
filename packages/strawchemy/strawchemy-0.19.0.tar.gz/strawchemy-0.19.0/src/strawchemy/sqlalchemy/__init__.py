"""SQLAlchemy integration for Strawchemy.

This package provides SQLAlchemy-based repository implementations for use with
Strawberry GraphQL.
"""

from __future__ import annotations

from .repository import SQLAlchemyGraphQLAsyncRepository, SQLAlchemyGraphQLRepository, SQLAlchemyGraphQLSyncRepository

__all__ = ("SQLAlchemyGraphQLAsyncRepository", "SQLAlchemyGraphQLRepository", "SQLAlchemyGraphQLSyncRepository")
