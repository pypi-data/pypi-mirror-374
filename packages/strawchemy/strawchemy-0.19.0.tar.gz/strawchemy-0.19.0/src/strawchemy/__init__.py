"""Custom DTO implementation."""

from __future__ import annotations

from .config.base import StrawchemyConfig
from .mapper import Strawchemy
from .sqlalchemy.hook import QueryHook
from .strawberry import ModelInstance
from .strawberry.mutation.input import Input
from .strawberry.mutation.types import (
    ErrorType,
    RequiredToManyUpdateInput,
    RequiredToOneInput,
    ToManyCreateInput,
    ToManyUpdateInput,
    ToOneInput,
    ValidationErrorType,
)
from .strawberry.repository import StrawchemyAsyncRepository, StrawchemySyncRepository
from .validation.base import InputValidationError

__all__ = (
    "ErrorType",
    "Input",
    "InputValidationError",
    "ModelInstance",
    "QueryHook",
    "RequiredToManyUpdateInput",
    "RequiredToOneInput",
    "Strawchemy",
    "StrawchemyAsyncRepository",
    "StrawchemyConfig",
    "StrawchemySyncRepository",
    "ToManyCreateInput",
    "ToManyUpdateInput",
    "ToOneInput",
    "ValidationErrorType",
)
