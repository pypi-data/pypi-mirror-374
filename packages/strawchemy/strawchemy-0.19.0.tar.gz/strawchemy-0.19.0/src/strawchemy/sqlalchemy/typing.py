from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal, TypeVar

from typing_extensions import TypeAlias

if TYPE_CHECKING:
    from collections import OrderedDict
    from collections.abc import Callable

    from sqlalchemy import Function
    from sqlalchemy.ext.asyncio import AsyncSession, async_scoped_session
    from sqlalchemy.orm import DeclarativeBase, Session, scoped_session
    from sqlalchemy.sql import SQLColumnExpression
    from strawchemy.strawberry.dto import OrderByEnum
    from strawchemy.strawberry.filters.base import GraphQLComparison

    from ._executor import QueryExecutor
    from .hook import QueryHook


__all__ = (
    "AnyAsyncSession",
    "AnySession",
    "AnySyncSession",
    "DeclarativeSubT",
    "DeclarativeT",
    "FunctionGenerator",
    "QueryExecutorT",
    "QueryHookCallable",
    "RelationshipSide",
    "SessionT",
    "StatementType",
)

DeclarativeT = TypeVar("DeclarativeT", bound="DeclarativeBase")
DeclarativeSubT = TypeVar("DeclarativeSubT", bound="DeclarativeBase")
QueryHookDeclarativeT = TypeVar("QueryHookDeclarativeT", bound="DeclarativeBase")
SessionT = TypeVar("SessionT", bound="AnySession")
QueryExecutorT = TypeVar("QueryExecutorT", bound="QueryExecutor[Any]")

RelationshipSide: TypeAlias = Literal["parent", "target"]
StatementType = Literal["lambda", "select"]
FunctionGenerator: TypeAlias = "Callable[..., Function[Any]]"
QueryHookCallable: TypeAlias = "QueryHook[QueryHookDeclarativeT]"
FilterMap: TypeAlias = "OrderedDict[tuple[type[Any], ...], type[GraphQLComparison]]"
AnySyncSession: TypeAlias = "Session | scoped_session[Session]"
AnyAsyncSession: TypeAlias = "AsyncSession | async_scoped_session[AsyncSession]"
AnySession: TypeAlias = "AnySyncSession | AnyAsyncSession"
OrderBySpec: TypeAlias = "tuple[SQLColumnExpression[Any], OrderByEnum]"
