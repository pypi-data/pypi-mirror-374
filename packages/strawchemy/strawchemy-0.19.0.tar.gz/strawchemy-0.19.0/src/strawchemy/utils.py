from __future__ import annotations

import inspect
import re
import sys
from typing import TYPE_CHECKING, Any, Optional, Union, get_args, get_origin

from strawchemy.typing import UNION_TYPES

if TYPE_CHECKING:
    from re import Pattern

__all__ = (
    "camel_to_snake",
    "is_type_hint_optional",
    "non_optional_type_hint",
    "snake_keys",
    "snake_to_camel",
    "snake_to_lower_camel_case",
)

_camel_to_snake_pattern: Pattern[str] = re.compile(r"((?<=[a-z0-9])[A-Z]|(?!^)(?<!_)[A-Z](?=[a-z]))")


def camel_to_snake(string: str) -> str:
    """Convert a camelcased string to snake case.

    See: https://stackoverflow.com/a/12867228
    """
    return _camel_to_snake_pattern.sub(r"_\1", string).lower()


def snake_to_camel(string: str) -> str:
    """Convert string to camel case.

    See: https://stackoverflow.com/a/19053800/10735573
    """
    return "".join(x.capitalize() for x in string.lower().split("_"))


def snake_to_lower_camel_case(snake_str: str) -> Any:
    """Convert string to lower camel case.

    See: https://stackoverflow.com/a/19053800/10735573
    """
    camel_string: str = snake_to_camel(snake_str)
    return snake_str[0].lower() + camel_string[1:]


def snake_keys(value: dict[str, Any]) -> dict[str, Any]:
    """Recursively convert dict keys to from camel case to snake case."""
    res: dict[Any, Any] = {}
    for k, v in value.items():
        to_snake: str = camel_to_snake(k)
        if isinstance(v, (list, tuple)):
            res[to_snake] = [snake_keys(el) for el in v]
        elif isinstance(v, dict):
            res[to_snake] = snake_keys(v)
        else:
            res[to_snake] = v
    return res


def non_optional_type_hint(type_hint: Any) -> Any:
    origin, args = get_origin(type_hint), get_args(type_hint)
    if origin is Optional:
        return args
    if origin in UNION_TYPES:
        union_args = tuple([arg for arg in args if arg not in (None, type(None))])
        if len(union_args) == 1:
            return union_args[0]
        return Union[union_args]
    return type_hint


def is_type_hint_optional(type_hint: Any) -> bool:
    """Whether the given type hint is considered as optional or not.

    Returns:
        `True` if arguments of the given type hint are optional

    Three cases are considered:
    ```
        Optional[str]
        Union[str, None]
        str | None
    ```
    In any other form, the type hint will not be considered as optional
    """
    origin = get_origin(type_hint)
    if origin is None:
        return False
    if origin is Optional:
        return True
    if origin in UNION_TYPES:
        args = get_args(type_hint)
        return any(arg is type(None) for arg in args)
    return False


def get_annotations(obj: Any) -> dict[str, Any]:
    """Get the annotations of the given object."""
    if sys.version_info >= (3, 10):
        return inspect.get_annotations(obj)
    return obj.__dict__.get("__annotations__", {}) if isinstance(obj, type) else getattr(obj, "__annotations__", {})
