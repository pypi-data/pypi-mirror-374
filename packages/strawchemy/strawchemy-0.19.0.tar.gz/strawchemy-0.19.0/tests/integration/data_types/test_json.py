from __future__ import annotations

from typing import TYPE_CHECKING, Any

import pytest

from sqlalchemy import Insert, MetaData, insert
from syrupy.assertion import SnapshotAssertion
from tests.integration.models import JSONModel, json_metadata
from tests.integration.types import mysql as mysql_types
from tests.integration.types import postgres as postgres_types
from tests.integration.types import sqlite as sqlite_types
from tests.integration.utils import to_graphql_representation
from tests.utils import maybe_async

if TYPE_CHECKING:
    from strawchemy.config.databases import DatabaseFeatures
    from strawchemy.typing import SupportedDialect

    from tests.integration.fixtures import QueryTracker
    from tests.integration.typing import RawRecordData
    from tests.typing import AnyQueryExecutor


@pytest.fixture
def metadata() -> MetaData:
    return json_metadata


@pytest.fixture
def seed_insert_statements(raw_json: RawRecordData) -> list[Insert]:
    return [insert(JSONModel).values(raw_json)]


@pytest.fixture
def async_query(dialect: SupportedDialect) -> type[Any]:
    if dialect == "postgresql":
        return postgres_types.JSONAsyncQuery
    if dialect == "mysql":
        return mysql_types.JSONAsyncQuery
    if dialect == "sqlite":
        return sqlite_types.JSONAsyncQuery
    pytest.skip(f"JSON tests can't be run on this dialect: {dialect}")


@pytest.fixture
def sync_query(dialect: SupportedDialect) -> type[Any]:
    if dialect == "postgresql":
        return postgres_types.JSONSyncQuery
    if dialect == "mysql":
        return mysql_types.JSONSyncQuery
    if dialect == "sqlite":
        return sqlite_types.JSONSyncQuery
    pytest.skip(f"JSON tests can't be run on this dialect: {dialect}")


# Tests for JSON-specific filters
@pytest.mark.parametrize(
    ("filter_name", "value", "expected_ids"),
    [
        pytest.param("contains", {"key1": "value1"}, [0], id="contains"),
        pytest.param(
            "containedIn",
            {"key1": "value1", "key2": 2, "key3": 3, "key4": None, "nested": {"inner": "value"}, "extra": "value"},
            [0, 2],
            id="containedIn",
        ),
        pytest.param("hasKey", "key1", [0], id="hasKey"),
        pytest.param("hasKeyAll", ["key1", "key2"], [0], id="hasKeyAll"),
        pytest.param("hasKeyAny", ["key1", "status"], [0, 1], id="hasKeyAny"),
    ],
)
@pytest.mark.snapshot
async def test_json_filters(
    filter_name: str,
    value: Any,
    expected_ids: list[int],
    any_query: AnyQueryExecutor,
    raw_json: RawRecordData,
    query_tracker: QueryTracker,
    sql_snapshot: SnapshotAssertion,
    db_features: DatabaseFeatures,
) -> None:
    if db_features.dialect == "sqlite" and filter_name in {"contains", "containedIn"}:
        pytest.skip(f"contains/containedIn not supported on {db_features.dialect}")
    if isinstance(value, list):
        value_str = ", ".join(to_graphql_representation(v, "input") for v in value)
        value_repr = f"[{value_str}]"
    else:
        value_repr = to_graphql_representation(value, "input")

    query = f"""
        {{
            json(filter: {{ dictCol: {{ {filter_name}: {value_repr} }} }}) {{
                id
                dictCol
            }}
        }}
    """
    result = await maybe_async(any_query(query))
    assert not result.errors
    assert result.data
    assert len(result.data["json"]) == len(expected_ids)

    for i, expected_id in enumerate(expected_ids):
        assert result.data["json"][i]["id"] == raw_json[expected_id]["id"]

    assert query_tracker.query_count == 1
    assert query_tracker[0].statement_formatted == sql_snapshot


@pytest.mark.snapshot
async def test_json_output(
    any_query: AnyQueryExecutor,
    raw_json: RawRecordData,
    query_tracker: QueryTracker,
    sql_snapshot: SnapshotAssertion,
) -> None:
    query = """
        {
            json {
                id
                dictCol
            }
        }
    """
    result = await maybe_async(any_query(query))
    assert not result.errors
    assert result.data

    for interval in result.data["json"]:
        expected_interval = next(f for f in raw_json if f["id"] == interval["id"])
        assert interval["dictCol"] == expected_interval["dict_col"]

    assert query_tracker.query_count == 1
    assert query_tracker[0].statement_formatted == sql_snapshot


@pytest.mark.parametrize(
    "path",
    [
        pytest.param("$.key1", id="key1"),
        pytest.param("$.key3", id="key3"),
        pytest.param("$.key4", id="key4"),
        pytest.param("$.nested", id="nested"),
    ],
)
@pytest.mark.snapshot
async def test_json_extract_path(
    path: str,
    any_query: AnyQueryExecutor,
    raw_json: RawRecordData,
    query_tracker: QueryTracker,
    sql_snapshot: SnapshotAssertion,
) -> None:
    query = f"""
        {{
            json {{
                id
                dictCol(path: "{path}")
            }}
        }}
    """
    result = await maybe_async(any_query(query))
    assert not result.errors
    assert result.data

    for json in result.data["json"]:
        expected_dict_col = next(f for f in raw_json if f["id"] == json["id"])
        assert json["dictCol"] == expected_dict_col["dict_col"].get(path.strip("$."), {})

    assert query_tracker.query_count == 1
    assert query_tracker[0].statement_formatted == sql_snapshot


@pytest.mark.snapshot
async def test_json_extract_inner_path(
    any_query: AnyQueryExecutor, raw_json: RawRecordData, query_tracker: QueryTracker, sql_snapshot: SnapshotAssertion
) -> None:
    query = """
        {
            json {
                id
                dictCol(path: "$.nested.inner")
            }
        }
    """
    result = await maybe_async(any_query(query))
    assert not result.errors
    assert result.data

    for json in result.data["json"]:
        expected_dict_col = next(f for f in raw_json if f["id"] == json["id"])
        assert json["dictCol"] == expected_dict_col["dict_col"].get("nested", {}).get("inner", {})

    assert query_tracker.query_count == 1
    assert query_tracker[0].statement_formatted == sql_snapshot
