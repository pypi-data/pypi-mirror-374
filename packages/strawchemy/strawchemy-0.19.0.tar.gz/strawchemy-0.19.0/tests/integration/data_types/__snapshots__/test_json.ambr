# serializer version: 1
# name: test_json_extract_inner_path[session-tracked-async-aiosqlite_engine]
  '''
  SELECT json_model.id,
         coalesce(json_model.dict_col -> ?, json_object()) AS json_model__dict_col
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_extract_inner_path[session-tracked-async-asyncmy_engine]
  '''
  SELECT json_model.id,
         coalesce(json_model.dict_col -> %s, json_object()) AS json_model__dict_col
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_extract_inner_path[session-tracked-async-asyncpg_engine]
  '''
  SELECT json_model.id,
         coalesce(jsonb_path_query_first(json_model.dict_col, CAST($1 AS JSONPATH)), CAST($2::JSONB AS JSONB)) AS json_model__dict_col
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_extract_inner_path[session-tracked-async-psycopg_async_engine]
  '''
  SELECT json_model.id,
         coalesce(jsonb_path_query_first(json_model.dict_col, CAST(%(param_1)s AS JSONPATH)), CAST(%(param_2)s::JSONB AS JSONB)) AS json_model__dict_col
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_extract_inner_path[session-tracked-sync-psycopg_engine]
  '''
  SELECT json_model.id,
         coalesce(jsonb_path_query_first(json_model.dict_col, CAST(%(param_1)s AS JSONPATH)), CAST(%(param_2)s::JSONB AS JSONB)) AS json_model__dict_col
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_extract_inner_path[session-tracked-sync-sqlite_engine]
  '''
  SELECT json_model.id,
         coalesce(json_model.dict_col -> ?, json_object()) AS json_model__dict_col
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_extract_path[session-tracked-async-asyncpg_engine]
  '''
  SELECT json_model.id,
         jsonb_path_query_first(json_model.dict_col, CAST($1 AS JSONPATH)) AS json_model__dict_col
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_extract_path[session-tracked-async-key1-aiosqlite_engine]
  '''
  SELECT json_model.id,
         coalesce(json_model.dict_col -> ?, json_object()) AS json_model__dict_col
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_extract_path[session-tracked-async-key1-asyncmy_engine]
  '''
  SELECT json_model.id,
         coalesce(json_model.dict_col -> %s, json_object()) AS json_model__dict_col
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_extract_path[session-tracked-async-key1-asyncpg_engine]
  '''
  SELECT json_model.id,
         coalesce(jsonb_path_query_first(json_model.dict_col, CAST($1 AS JSONPATH)), CAST($2::JSONB AS JSONB)) AS json_model__dict_col
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_extract_path[session-tracked-async-key1-psycopg_async_engine]
  '''
  SELECT json_model.id,
         coalesce(jsonb_path_query_first(json_model.dict_col, CAST(%(param_1)s AS JSONPATH)), CAST(%(param_2)s::JSONB AS JSONB)) AS json_model__dict_col
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_extract_path[session-tracked-async-key3-aiosqlite_engine]
  '''
  SELECT json_model.id,
         coalesce(json_model.dict_col -> ?, json_object()) AS json_model__dict_col
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_extract_path[session-tracked-async-key3-asyncmy_engine]
  '''
  SELECT json_model.id,
         coalesce(json_model.dict_col -> %s, json_object()) AS json_model__dict_col
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_extract_path[session-tracked-async-key3-asyncpg_engine]
  '''
  SELECT json_model.id,
         coalesce(jsonb_path_query_first(json_model.dict_col, CAST($1 AS JSONPATH)), CAST($2::JSONB AS JSONB)) AS json_model__dict_col
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_extract_path[session-tracked-async-key3-psycopg_async_engine]
  '''
  SELECT json_model.id,
         coalesce(jsonb_path_query_first(json_model.dict_col, CAST(%(param_1)s AS JSONPATH)), CAST(%(param_2)s::JSONB AS JSONB)) AS json_model__dict_col
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_extract_path[session-tracked-async-key4-aiosqlite_engine]
  '''
  SELECT json_model.id,
         coalesce(json_model.dict_col -> ?, json_object()) AS json_model__dict_col
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_extract_path[session-tracked-async-key4-asyncmy_engine]
  '''
  SELECT json_model.id,
         coalesce(json_model.dict_col -> %s, json_object()) AS json_model__dict_col
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_extract_path[session-tracked-async-key4-asyncpg_engine]
  '''
  SELECT json_model.id,
         coalesce(jsonb_path_query_first(json_model.dict_col, CAST($1 AS JSONPATH)), CAST($2::JSONB AS JSONB)) AS json_model__dict_col
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_extract_path[session-tracked-async-key4-psycopg_async_engine]
  '''
  SELECT json_model.id,
         coalesce(jsonb_path_query_first(json_model.dict_col, CAST(%(param_1)s AS JSONPATH)), CAST(%(param_2)s::JSONB AS JSONB)) AS json_model__dict_col
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_extract_path[session-tracked-async-nested-aiosqlite_engine]
  '''
  SELECT json_model.id,
         coalesce(json_model.dict_col -> ?, json_object()) AS json_model__dict_col
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_extract_path[session-tracked-async-nested-asyncmy_engine]
  '''
  SELECT json_model.id,
         coalesce(json_model.dict_col -> %s, json_object()) AS json_model__dict_col
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_extract_path[session-tracked-async-nested-asyncpg_engine]
  '''
  SELECT json_model.id,
         coalesce(jsonb_path_query_first(json_model.dict_col, CAST($1 AS JSONPATH)), CAST($2::JSONB AS JSONB)) AS json_model__dict_col
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_extract_path[session-tracked-async-nested-psycopg_async_engine]
  '''
  SELECT json_model.id,
         coalesce(jsonb_path_query_first(json_model.dict_col, CAST(%(param_1)s AS JSONPATH)), CAST(%(param_2)s::JSONB AS JSONB)) AS json_model__dict_col
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_extract_path[session-tracked-async-psycopg_async_engine]
  '''
  SELECT json_model.id,
         jsonb_path_query_first(json_model.dict_col, CAST(%(param_1)s AS JSONPATH)) AS json_model__dict_col
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_extract_path[session-tracked-sync-key1-psycopg_engine]
  '''
  SELECT json_model.id,
         coalesce(jsonb_path_query_first(json_model.dict_col, CAST(%(param_1)s AS JSONPATH)), CAST(%(param_2)s::JSONB AS JSONB)) AS json_model__dict_col
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_extract_path[session-tracked-sync-key1-sqlite_engine]
  '''
  SELECT json_model.id,
         coalesce(json_model.dict_col -> ?, json_object()) AS json_model__dict_col
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_extract_path[session-tracked-sync-key3-psycopg_engine]
  '''
  SELECT json_model.id,
         coalesce(jsonb_path_query_first(json_model.dict_col, CAST(%(param_1)s AS JSONPATH)), CAST(%(param_2)s::JSONB AS JSONB)) AS json_model__dict_col
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_extract_path[session-tracked-sync-key3-sqlite_engine]
  '''
  SELECT json_model.id,
         coalesce(json_model.dict_col -> ?, json_object()) AS json_model__dict_col
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_extract_path[session-tracked-sync-key4-psycopg_engine]
  '''
  SELECT json_model.id,
         coalesce(jsonb_path_query_first(json_model.dict_col, CAST(%(param_1)s AS JSONPATH)), CAST(%(param_2)s::JSONB AS JSONB)) AS json_model__dict_col
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_extract_path[session-tracked-sync-key4-sqlite_engine]
  '''
  SELECT json_model.id,
         coalesce(json_model.dict_col -> ?, json_object()) AS json_model__dict_col
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_extract_path[session-tracked-sync-nested-psycopg_engine]
  '''
  SELECT json_model.id,
         coalesce(jsonb_path_query_first(json_model.dict_col, CAST(%(param_1)s AS JSONPATH)), CAST(%(param_2)s::JSONB AS JSONB)) AS json_model__dict_col
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_extract_path[session-tracked-sync-nested-sqlite_engine]
  '''
  SELECT json_model.id,
         coalesce(json_model.dict_col -> ?, json_object()) AS json_model__dict_col
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_extract_path[session-tracked-sync-psycopg_engine]
  '''
  SELECT json_model.id,
         jsonb_path_query_first(json_model.dict_col, CAST(%(param_1)s AS JSONPATH)) AS json_model__dict_col
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-async-containedIn-asyncmy_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_contains(CAST(%s AS JSON), json_model.dict_col)
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-async-containedIn-asyncpg_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_model.dict_col <@ $1::
   JSONB
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-async-containedIn-psycopg_async_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_model.dict_col <@ %(param_1)s::
   JSONB
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-async-contains-asyncmy_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_contains(json_model.dict_col, CAST(%s AS JSON))
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-async-contains-asyncpg_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_model.dict_col @> $1::
   JSONB
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-async-contains-psycopg_async_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_model.dict_col @> %(param_1)s::
   JSONB
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-async-hasKey-aiosqlite_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_extract(json_model.dict_col, ?) IS NOT NULL
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-async-hasKey-asyncmy_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_contains_path(json_model.dict_col, %s, %s)
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-async-hasKey-asyncpg_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_model.dict_col ? $1::VARCHAR
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-async-hasKey-psycopg_async_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_model.dict_col ? %(param_1)s::VARCHAR
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-async-hasKeyAll-aiosqlite_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_extract(json_model.dict_col, ?) IS NOT NULL
     AND json_extract(json_model.dict_col, ?) IS NOT NULL
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-async-hasKeyAll-asyncmy_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_contains_path(json_model.dict_col, %s, %s, %s)
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-async-hasKeyAll-asyncpg_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_model.dict_col ? & CAST($1::TEXT[] AS TEXT[])
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-async-hasKeyAll-psycopg_async_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_model.dict_col ? & CAST(%(param_1)s::TEXT[] AS TEXT[])
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-async-hasKeyAny-aiosqlite_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_extract(json_model.dict_col, ?) IS NOT NULL
      OR json_extract(json_model.dict_col, ?) IS NOT NULL
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-async-hasKeyAny-asyncmy_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_contains_path(json_model.dict_col, %s, %s, %s)
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-async-hasKeyAny-asyncpg_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_model.dict_col ? | CAST($1::TEXT[] AS TEXT[])
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-async-hasKeyAny-psycopg_async_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_model.dict_col ? | CAST(%(param_1)s::TEXT[] AS TEXT[])
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-sync-containedIn-psycopg_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_model.dict_col <@ %(param_1)s::
   JSONB
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-sync-contains-psycopg_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_model.dict_col @> %(param_1)s::
   JSONB
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-sync-hasKey-psycopg_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_model.dict_col ? %(param_1)s::VARCHAR
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-sync-hasKey-sqlite_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_extract(json_model.dict_col, ?) IS NOT NULL
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-sync-hasKeyAll-psycopg_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_model.dict_col ? & CAST(%(param_1)s::TEXT[] AS TEXT[])
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-sync-hasKeyAll-sqlite_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_extract(json_model.dict_col, ?) IS NOT NULL
     AND json_extract(json_model.dict_col, ?) IS NOT NULL
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-sync-hasKeyAny-psycopg_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_model.dict_col ? | CAST(%(param_1)s::TEXT[] AS TEXT[])
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_filters[session-tracked-sync-hasKeyAny-sqlite_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   WHERE json_extract(json_model.dict_col, ?) IS NOT NULL
      OR json_extract(json_model.dict_col, ?) IS NOT NULL
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_output[session-tracked-async-aiosqlite_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_output[session-tracked-async-asyncmy_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_output[session-tracked-async-asyncpg_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_output[session-tracked-async-psycopg_async_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_output[session-tracked-sync-psycopg_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
# name: test_json_output[session-tracked-sync-sqlite_engine]
  '''
  SELECT json_model.dict_col,
         json_model.id
    FROM json_model AS json_model
   ORDER BY json_model.id ASC
  '''
# ---
